### ISSUE TITLE
Circuit breaking not working when there are no endpoints available

### ISSUE THREAD

**Issue by Nbarola** [linkerd/linkerd2#11580]
### What is the issue?

When there are no endpoints available then the circuit breaking is not triggered

### How can it be reproduced?

Scale the replicas to 0 for the deployment where circuit breaking is applied

### Logs, error output, etc

No message regarding the circuit being closed in the logs of the caller service

EDIT:-
Also, in the prometheus was looking at this metrics

```
outbound_http_balancer_endpoints{app="app-name", endpoint_state="pending"}
```
In case when the circuit breaking is working we can see the endpoint error-injector in pending state but when we have no replicas then we don't see it in pending state. 

### output of `linkerd check -o short`

```
linkerd-version
---------------
‼ cli is up-to-date
    is running version 2.13.6 but the latest stable version is 2.14.2
    see https://linkerd.io/2.13/checks/#l5d-version-cli for hints

control-plane-version
---------------------
‼ control plane is up-to-date
    is running version 2.13.6 but the latest stable version is 2.14.2
    see https://linkerd.io/2.13/checks/#l5d-version-control for hints

linkerd-control-plane-proxy
---------------------------
‼ control plane proxies are up-to-date
    some proxies are not running the current version:
	* linkerd-destination-d7b9f5cf9-2v24g (stable-2.13.6)
	* linkerd-destination-d7b9f5cf9-b6dpm (stable-2.13.6)
	* linkerd-destination-d7b9f5cf9-tpf76 (stable-2.13.6)
	* linkerd-identity-745c659c7b-2z265 (stable-2.13.6)
	* linkerd-identity-745c659c7b-nljc4 (stable-2.13.6)
	* linkerd-identity-745c659c7b-pj4v6 (stable-2.13.6)
	* linkerd-proxy-injector-5b757bc787-66qqj (stable-2.13.6)
	* linkerd-proxy-injector-5b757bc787-hpx52 (stable-2.13.6)
	* linkerd-proxy-injector-5b757bc787-xnkpg (stable-2.13.6)
    see https://linkerd.io/2.13/checks/#l5d-cp-proxy-version for hints

linkerd-viz
-----------
‼ viz extension proxies are up-to-date
    some proxies are not running the current version:
	* metrics-api-746ccdc44b-8j5nr (stable-2.13.6)
	* metrics-api-746ccdc44b-cn8bg (stable-2.13.6)
	* metrics-api-746ccdc44b-v64gn (stable-2.13.6)
	* prometheus-7b47989794-z7bgr (stable-2.13.6)
	* tap-66dc4d6ccf-8pj7x (stable-2.13.6)
	* tap-66dc4d6ccf-dlgpg (stable-2.13.6)
	* tap-66dc4d6ccf-v72c7 (stable-2.13.6)
	* tap-injector-ffdf8c777-qlnzc (stable-2.13.6)
	* web-75d5ff89b6-72z6n (stable-2.13.6)
	* web-75d5ff89b6-htcll (stable-2.13.6)
	* web-75d5ff89b6-pkv7b (stable-2.13.6)
    see https://linkerd.io/2.13/checks/#l5d-viz-proxy-cp-version for hints

Status check results are √

```

### Environment

- EKS 1.24
- Host OS - linux
- Linkerd version 2.13.6

### Would you like to work on fixing this bug?

None

**Comment by DavidMcLaughlin** [comment]
Can you explain what the desired behavior is? Having a hard time understanding if this is a metrics issue or something else.

**Comment by Nbarola** [comment]
Hey @DavidMcLaughlin, with the Circuit breaking, we expect the same behavior when there are no active downstream endpoints. Typically in a normal scenario, the circuit gets tripped when the downstream endpoint returns 5XX errors and the upstream then stops sending the request to those endpoints, and if all the downstream endpoints start returning 5XX then the upstream stops sending any request. So, when there are no endpoints (lets say the downstream pods are down) then also the circuit gets tripped as when there are no endpoints available then we see 503 service unavailable errors in the proxy logs, so the upstream should stop sending the request and only resume when the downstream is up again.

**Comment by kflynn** [comment]
@Nbarola Sorry for the delay here! I had to understand a lot more about our circuit breaking implementation than I did at first -- but, ultimately, I feel like this one doesn't really sound like an issue with Linkerd circuit breaking per se. You're describing exactly what _should_ happen, from our point of view: once all the backends are unavailable, Linkerd responds to client requests with an immediate 503, _because_ there are no backends available.

However, Linkerd has no control over the client's actions. If you want the client to stop sending requests at that point, that will need to be implemented in the client. One important note is that when all the circuits are broken, the 503 comes from Linkerd itself, not from any backend -- so the client can keep sending without actually causing extra network traffic. And honestly, Linkerd _needs_ the client to send requests every so often to allow the circuit breaking probes to work, so that it can allow the broken circuits to close again -- but, of course, the client could decide to slow down.

So overall, I think you're probably seeing intended behavior here -- does this explanation make sense?
